src/libsyntax_ext/deriving/hash.rs

```diff
@@ -77,15 +81,18 @@ fn hash_substructure(cx: &mut ExtCtxt, trait_span: Span, substr: &Substructure)
 
     let fields = match *substr.fields {
         Struct(_, ref fs) => fs,
-        EnumMatching(index, variant, ref fs) => {
-            // Determine the discriminant. We will feed this value to the byte
-            // iteration function.
-            let discriminant = match variant.node.disr_expr {
-                Some(ref d) => d.clone(),
-                None => cx.expr_usize(trait_span, index)
-            };
+        EnumMatching(_, _, ref fs) => {
+            let path = cx.std_path(&["intrinsics", "discriminant_value"]);
```

Comment:
Yeah, this should be factored out into a function (also used in `#[derive(PartialOrd)]`)